File: src/main/java/io/pageobject/generator/locator/protractor/ButtonTextLocatorSource.java
package io.pageobject.generator.locator.protractor;

import io.pageobject.generator.GeneratorContext;
import io.pageobject.generator.locator.AbstractLocatorSource;
import org.jsoup.nodes.Element;

import static com.google.common.base.Strings.isNullOrEmpty;
import static io.pageobject.generator.Expressions.hasExpression;

public class ButtonTextLocatorSource extends AbstractLocatorSource {

    @Override
    public String extractLocatorValue(Element element, GeneratorContext context) {
        String text = element.text();
        String locator = null;

        if (isNullOrEmpty(text)) {
            locator = element.attr("value");
        } else if (!hasExpression(text)) {
            locator = text;
        }

        return locator;
    }

    public String locatorTemplate() {
        return "by.buttonText('%s')";
    }

}

File: lib/locator/protractor/button-text-locator-source.js
"use strict";

const AbstractLocatorSource = require('../abstract-locator-source');
const text = require('../../dom-helpers').text;
const getAttribute = require('../../dom-helpers').getAttribute;
const hasExpression = require('../../expressions').hasExpression;
const _ = require('lodash');

const template = _.template("by.buttonText('<%= value %>')");

class ButtonTextLocatorSource extends AbstractLocatorSource {

  extractLocatorValue(context) {
    let element = context.element;
    let content = text(element);
    let locator = null;

    if (_.isEmpty(content)) {
      locator = getAttribute(element, 'value');
    } else if (!hasExpression(content)) {
      locator = content;
    }

    return locator;
  }

  locatorTemplate() {
    return template;
  }

}

module.exports = ButtonTextLocatorSource;